Add docs for TCP apps access (#14978)
gravitational/teleport: docs/pages/application-access/guides/tcp.mdx

---
title: TCP Application Access (Preview)
description: How to configure Teleport for accessing plain TCP apps
---

Teleport can provide access to any TCP-based application. This allows users to
connect to applications which Teleport doesn't natively support such as SMTP
servers or databases not yet natively supported in Database Access.

## Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

(!docs/pages/includes/tctl.mdx!)

- TCP application to connect to. In this guide we'll use a PostgreSQL running

>   in Docker as an example, you can also use any TCP-based application you may

Reviewer: ```suggestion
  in Docker as an example. You can also use any TCP-based application you may
```

  already have.
- Host where you will run the Teleport Application Service.

We will assume your Teleport cluster is accessible at `teleport.example.com`
and `*.teleport.example.com`. You can substitute the address of your Teleport
Proxy Service. (For Teleport Cloud customers, this will be similar to
`mytenant.teleport.sh`.)

> <Admonition type="note" title="Application Access and DNS">

Reviewer: ```suggestion
<Admonition type="note" title="Application Access and DNS" scope={["oss", "enterprise"]} scopeOnly>
```
So Cloud users don't create unnecessary DNS records

(!docs/pages/includes/dns-app-access.mdx!)
</Admonition>

## Step 1/4. Start PostgreSQL container

Skip this step if you already have an application you'd like to connect to.

Start a PostgreSQL server in a Docker container:

```code
$ docker run --name postgres -p 5432:5432 -e POSTGRES_PASSWORD=<pass> -d postgres
```

## Step 2/4. Start Teleport Application Service

Teleport Application Service requires a valid auth token to join the cluster.

<ScopedBlock scope={["oss", "enterprise"]}>
To generate one, run the following command on your Auth Service node:

```code
$ tctl tokens add --type=app
```

Next, create a Teleport user with the `access` role that will allow it to
connect to cluster applications:

```code
$ tctl users add --roles=access alice
```
</ScopedBlock>

<ScopedBlock scope={["cloud"]}>
To generate one, log into your Cloud tenant and run the following command:

```code
$ tsh login --proxy=mytenant.teleport.sh
$ tctl tokens add --type=app
```
</ScopedBlock>

Save the generated token in `/tmp/token` on the node where Application Service
will run.

> Now, install Teleport on Application Service node. It must be able to reach

Reviewer: ```suggestion
Now, install Teleport on the Application Service node. It must be able to reach
```

```yaml
teleport:
  auth_token: "/tmp/token"
  auth_servers:
  - teleport.example.com:3080
auth_service:
  enabled: "no"
ssh_service:
  enabled: "no"
proxy_service:
  enabled: "no"
app_service:
  enabled: "yes"
  apps:
  - name: "tcp-app"
    uri: tcp://localhost:5432
```

Note that the URI scheme must be `tcp://` in order for Teleport to recognize
this as a TCP application.

Start Teleport:

```code
$ teleport start
```

## Step 3/4. Start app proxy

Log into your Teleport cluster and view available applications:

```code
$ tsh login --proxy=teleport.example.com
$ tsh app ls
Application Description   Type Public Address                   Labels
----------- ------------- ---- -------------------------------- -----------
tcp-app                   TCP  tcp-app.root.gravitational.io
```

Your TCP application should show up and be denoted with a `TCP` type.

Now log into the application:

```code
$ tsh app login tcp-app
Logged into TCP app tcp-app. Start the local TCP proxy for it:

  tsh proxy app tcp-app

Then connect to the application through this proxy.
```

Next, start a local proxy for it:

```code
$ tsh proxy app tcp-app
Proxying connections to tcp-app on 127.0.0.1:55868
```

> The local proxy will setup a local listener that will proxy all connection to

Reviewer: ```suggestion
The `tsh proxy app` command will set up a listener that will proxy all connections to
```

---

> title: Use Teleport for accessing plain TCP apps

Reviewer: We use title case for titles (sentence case for headings)

description: How to configure Teleport for accessing plain TCP apps.
---

> # TCP Apps Access (Preview)

Reviewer: Same comment as above re: the H1

Teleport can provide access to any TCP-based application. This allows users to
connect to applications which Teleport doesn't natively support such as SMTP
servers or databases not yet natively supported in Database Access.

## Prerequisites

> You will need a running Teleport cluster, either self-hosted or in Teleport Cloud.

Reviewer: I think we should use a similar format as the Application Access Getting Started guide for the Prerequisites (http://localhost:3000/docs/application-access/getting-started/#prerequisites), with a separate step for installing and starting Teleport on the host where the App Service will run. It doesn't add much length to the guide and saves the user from needing to switch context. 

This will be especially useful for cases where a user is only interested in plain TCP apps, not web applications.

We'll assume that you followed the [Getting Started](../getting-started.mdx)
guide or the general [App Access Usage](./connecting-apps.mdx) guide to install
Teleport and setup an Application Service.

## Step 1/3. Configure TCP app

Include your TCP application in the `teleport.yaml` configuration of Teleport
Application Service:

```yaml
app_service:
  enabled: "yes"
  apps:
  - name: "tcp-app"
    uri: tcp://localhost:9876
    labels:
      app: tcp-app
```

Note that the URI scheme must be `tcp://` in order for Teleport to recognize
this as a TCP application.

> You can also register the application [dynamically](./dynamic-registration.mdx)

Reviewer: I think we can either remove this sentence or add it to a Details box that's closed by default. For this sentence to be meaningful, the user needs to visit the Dynamic Registration guide, read through it, and work out how to convert the instructions above into a dynamic registration, taking them away from the flow of this guide.

using `tctl create`.

## Step 2/3. Start app proxy

Log into your Teleport cluster and view available applications:

```code
$ tsh login --proxy=teleport.example.com
$ tsh app ls
Application Description   Type Public Address                   Labels
----------- ------------- ---- -------------------------------- -----------
tcp-app                   TCP  tcp-app.root.gravitational.io    app=tcp-app
```

Your TCP application should show up and be denoted with a `TCP` type.

Now log into the application:

```code
$ tsh app login tcp-app
Logged into TCP app tcp-app. Start the local TCP proxy for it:

  tsh proxy app tcp-app

Then connect to the application through this proxy.
```

> Before you can connect to the application, you need to start a local proxy for

Reviewer: If this step comes before the previous command, we should put it first. Otherwise, a user running each command in sequence will hit an error at `tsh app login`.

it:

> ```code

Reviewer: I think it would be good to add a sentence explaining what the local proxy does.

$ tsh proxy app tcp-app
Proxying connections to tcp-app on 127.0.0.1:55868
```

## Step 3/3. Connect

> Once the local proxy is running, you can connect to the application through

Reviewer: This is the first mention of a tunnel in the guide. I think we should either explain what "the tunnel" refers to or remove "through the tunnel".
